lenskit-core/src/main/java/org/lenskit/data/dao/file/DelimitedColumnEntityFormat.java

```diff
@@ -0,0 +1,235 @@
+        readHeader = header;
+        headerLines = 1;
+    }
+
+    /**
+     * Get the number of header lines to read.
+     * @return The number of header lines to read.
+     */
+    public int getHeaderLines() {
+        return headerLines;
+    }
+
+    /**
+     * Set the number of header lines to read.  Setting this **disables** {@link #setHeader(boolean)}.
+     * @param lines The number of header lines to read.
+     */
+    public void setHeaderLines(int lines) {
+        headerLines = lines;
+        readHeader = false;
+    }
+
+    /**
+     * Query whether this format uses the header line(s).
+     * @return `true` if the reader will parse a header line.
+     */
+    boolean usesHeader() {
+        return readHeader;
+    }
+
+    /**
+     * Set the entity type.
+     * @param type The entity type.
+     */
+    public void setEntityType(EntityType type) {
+        entityType = type;
+    }
+
+    /**
+     * Get the entity type.
+     * @return The entity type.
+     */
+    @Override
+    public EntityType getEntityType() {
+        return entityType;
+    }
+
+    /**
+     * Set the entity builder class.
+     * @param builder The entity builder class.
+     */
+    public void setEntityBuilder(Class<? extends EntityBuilder> builder) {
+        entityBuilder = builder;
+    }
+
+    /**
+     * Get the entity builder class.
+     * @return The entity builder class.
+     */
+    public Class<? extends EntityBuilder> getEntityBuilder() {
+        return entityBuilder;
+    }
+
+    /**
+     * Instantiate a new entity builder.
+     * @return A new entity builder.
+     */
+    public EntityBuilder newEntityBuilder() {
+        if (entityBuilderCtor == null || !entityBuilderCtor.getDeclaringClass().equals(entityBuilder)) {
+            try {
+                entityBuilderCtor = entityBuilder.getConstructor(EntityType.class);
+            } catch (NoSuchMethodException e) {
+                throw new IllegalArgumentException("cannot find suitable constructor for " + entityBuilder);
+            }
+        }
+        try {
+            return entityBuilderCtor.newInstance(entityType);
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new RuntimeException("could not instantiate entity builder", e);
+        }
+    }
+
+    /**
+     * Add a column.
+     * @param attr The attribute to add as a column, or `null` to skip the next column.
+     * @throws IllegalStateException if non-labeled columns have already been specified.
+     */
+    public void addColumn(@Nullable TypedName<?> attr) {
+        if (columns == null) {
+            Preconditions.checkState(labeledColumns == null, "mixed labeled and unlabeled columns");
+            columns = new ArrayList<>();
+        }
+        columns.add(attr);
+    }
+
+    /**
+     * Add columns to a format.
+     * @param columns The columns to add.
+     */
+    public void addColumns(TypedName<?>... columns) {
```

Comment:
I'm not sure if I want to allow this to be used if the other ordered column add is used... At the vary least it needs a verbose comment like the past one.